apis/stacks/v1alpha1/stackdefinition.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StackDefinitionSpec defines the desired state of StackDefinition
+type StackDefinitionSpec struct {
+	AppMetadataSpec `json:",inline"`
+	CRDs            CRDList                 `json:"customresourcedefinitions,omitempty"`
+	Controller      ControllerSpec          `json:"controller,omitempty"`
```

Comment:
Oh okay, I think you're consolidating `stack.yaml` work as you outlined [here](https://github.com/crossplaneio/crossplane/issues/1209#issuecomment-579568139) I expect that controller would have a logic like `if Spec.Behavior is not empty, ignore Spec.Controller`, right? Maybe you'd need to make type of `Behavior` a pointer type.

Reply:
> I expect that controller would have a logic like if Spec.Behavior is not empty, ignore Spec.Controller

Yeah, something like that. Our options for if both a controller and behavior are specified would be:

* Use the specified controller
* Throw away the specified controller and use the one generated from the behaviors instead

And I would lean toward the second one, since behaviors aren't as useful if a custom controller is used (unless that controller looks at the behaviors, which could also be a valid use-case eventually). And because it's simpler to conceptualize.

> Maybe you'd need to make type of Behavior a pointer type

I had a noob question and was wondering why we would do that, but I did some googling, and [go's json explains](https://golang.org/pkg/encoding/json/#Marshal):

```
The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as . . . a nil pointer, . . .
```

I believe I agree with the pattern of omitting empty optional fields by default, so yeah, this seems like a good idea.

Reply:
Giving it a second thought, I think we could leave it if the user specified it. It's a normal case where user would want to pin down the version of the templating engine image by copying the `Controller` section from a live cluster.

Reply:
That sounds reasonable to me. Though, is that the way we want to give them control over the version? Trying to think through it right now. This seems reasonable too, because it gives them direct control over the lower-level parameters without adding a (potentially imperfect) abstraction in between.

Reply:
> This seems reasonable too, because it gives them direct control over the lower-level parameters without adding a (potentially imperfect) abstraction in between.

I agree. I believe that's a general pattern in Kubernetes. For example, `Deployment` allows you to specify full pod via `PodTemplateSpec` and you fill only the ones you'd like to override the default.